app/org/sagebionetworks/bridge/services/ReportService.java

```diff
@@ -0,0 +1,113 @@
+            LocalDate startDate, LocalDate endDate) {
+        // ReportDataKey validates all parameters to this method
+        
+        startDate = defaultValueToYesterday(startDate);
+        endDate = defaultValueToYesterday(endDate);
+        validateDateRange(startDate, endDate);
+
+        ReportDataKey key = makeKey(null, identifier, studyId, ReportType.STUDY);
+        return reportDataDao.getReportData(key, startDate, endDate);
+    }
+    
+    public DateRangeResourceList<? extends ReportData> getParticipantReport(StudyIdentifier studyId, String identifier, String healthCode, LocalDate startDate, LocalDate endDate) {
+        // ReportDataKey validates all parameters to this method
+        
+        startDate = defaultValueToYesterday(startDate);
+        endDate = defaultValueToYesterday(endDate);
+        validateDateRange(startDate, endDate);
+
+        ReportDataKey key = makeKey(healthCode, identifier, studyId, ReportType.PARTICIPANT);
+        return reportDataDao.getReportData(key, startDate, endDate);
+    }
+
+    public void saveStudyReport(StudyIdentifier studyId, String identifier, ReportData reportData) {
+        checkNotNull(reportData);
+        // ReportDataKey validates all other parameters to this method
+
+        ReportDataKey key = makeKey(null, identifier, studyId, ReportType.STUDY);
+        reportData.setKey(key.getKeyString());
+        
+        reportDataDao.saveReportData(reportData);
+    }
+    
+    public void saveParticipantReport(StudyIdentifier studyId, String identifier, String healthCode, ReportData reportData) {
+        checkNotNull(reportData);
+        // ReportDataKey validates all other parameters to this method
+        
+        ReportDataKey key = makeKey(healthCode, identifier, studyId, ReportType.PARTICIPANT);
+        reportData.setKey(key.getKeyString());
+        
+        reportDataDao.saveReportData(reportData);
+    }
+    
+    public void deleteStudyReport(StudyIdentifier studyId, String identifier) {
+        // ReportDataKey validates all parameters to this method
+
+        ReportDataKey key = makeKey(null, identifier, studyId, ReportType.STUDY);
+        
+        reportDataDao.deleteReportData(key);
+    }
+    
+    public void deleteParticipantReport(StudyIdentifier studyId, String identifier, String healthCode) {
+        // ReportDataKey validates all parameters to this method
+        
+        ReportDataKey key = makeKey(healthCode, identifier, studyId, ReportType.PARTICIPANT);
+        
+        reportDataDao.deleteReportData(key);
+    }
+    
+    private ReportDataKey makeKey(String healthCode, String identifier, StudyIdentifier studyId, ReportType reportType) {
+        return new ReportDataKey.Builder()
+                .withHealthCode(healthCode)
+                .withReportType(reportType)
+                .withIdentifier(identifier)
+                .withStudyIdentifier(studyId).build();
+    }
```

Comment:
This defeats the whole purpose of using the builder pattern. The point of the builder pattern is that when we add a new parameter, we don't have to update N^2 constructors and constructor usages. If we anticipate adding new keys (which I suspect is likely), we should remove this faux-constructor and just use the builder directly. If we are sure we'll never add new report keys, then we should get rid of the builder and just use the constructor directly.